### DISCUSSION TITLE
Projects Embedding Janet?

### DISCUSSION THREAD

**Discussion by ghost** [janet-lang/janet#1396] [General]
(**General** doesn't seem right, but all the others are definitely wrong.)

What are some open-source projects embedding Janet for extension/scripting? I can find projects *written in* Janet easily enough, but projects merely *using* it are harder to come by. I'd like to see how that's done in practice with an in-house API, as opposed to all the example code I could find which just uses the default environment and nothing else.

**Comment by sogaiu** [comment]
May be some of what's listed [here](https://github.com/sogaiu/janet-features-demos?tab=readme-ov-file#add-to-a-project-with-just-janetc-and-janeth) is sort of along the lines of what you're looking for?

**Reply by ghost** [reply]
This is helpful, thanks. It seems the Lua pattern of one global API object with many sub-namespaces doesn't really work. I didn't realise I'd fallen down the space tree; I guess I need a swim.

**Comment by iacore** [comment]
I have some exploratory projects that are written in Zig.

https://www.1a-insec.net/blog/16-buy-janet-get-gc-free/
https://www.1a-insec.net/blog/18-janet-as-database/
https://www.1a-insec.net/blog/20-janet-graphical-repl/

**Comment by turbohz** [comment]
TIC-80 might be worth a look.
https://github.com/nesbox/TIC-80

**Reply by ghost** [reply]
Wow, that is many more languages than I expected it to run. I thought it stored code as Lua text? I guess I just assumed since that's what PICO8 does. I also assumed closed was really closed, but they make it available with a bit of DIY; it always rubbed me the wrong way that P8 is closed-source when users are expected to distribute their creations for free.

Anyways that's a big sprawling project. Could you please point out the C file that sets up the VM entry point?

**Reply by sogaiu** [reply]
The `initJanet` function of [this file](https://github.com/nesbox/TIC-80/blob/main/src/api/janet.c) calls `janet_init`.

@AlecTroemel would know better though :)

**Reply by AlecTroemel** [reply]
yep that's the file! 

specifically [this function here](https://github.com/nesbox/TIC-80/blob/main/src/api/janet.c#L1059) is where the 
- a Janet environment is created. I believe this stores all the state of the running VM _between_ iterations. stuff like variables and functions.
- assigns a global [GameFiber](https://github.com/nesbox/TIC-80/blob/main/src/api/janet.c#L1059Game) variable to a janet fiber. I believe a fiber encapsulates ticks or iterations of the janet VM.. at least thats how I think about it in tic80's context.
- registers the TIC-80 API, which are all the binding function that take up a bulk of the file
- redirects logs and errors, since tic80 has its own "terminal" 

Another unique thing about the TIC-80 setup is that it expects the janet source to include a `TIC` function. That function is [grabbed  out of the vm and called](https://github.com/nesbox/TIC-80/blob/main/src/api/janet.c#L1059Game) every frame. Other use cases may not require that :laughing:  

(someone please correct me if I'm wrong about fibers/environments!)
